File: src/codexpro_bridge/capabilities/registry.py
"""Capability-module registry and public tool-surface governance."""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Any, Callable

from codexpro_bridge.core.errors import BridgeError

Health = Callable[[], dict[str, Any]]
Shutdown = Callable[[], None]
Install = Callable[[Any], None]


@dataclass(frozen=True, slots=True)
class ModuleDescriptor:
    """Metadata and installer for one independently controlled capability module."""

    module_id: str
    version: str
    tools: tuple[str, ...]
    health: Health
    install: Install
    shutdown: Shutdown | None = None


class CapabilityRegistry:
    """Own module selection, mounting, collision checks and residue detection."""

    def __init__(self, enabled_modules: tuple[str, ...] | None = None) -> None:
        self._modules: dict[str, ModuleDescriptor] = {}
        self._tool_owner: dict[str, str] = {}
        self._enabled_selector = None if enabled_modules is None else tuple(dict.fromkeys(enabled_modules))
        self._tool_source: Callable[[], set[str]] | None = None

    def register(self, descriptor: ModuleDescriptor) -> None:
        if not descriptor.module_id or descriptor.module_id in self._modules:
            raise BridgeError("registry_collision", "Bridge capability module is duplicated")
        if not descriptor.tools:
            raise BridgeError("registry_invalid", "Bridge capability module must declare at least one public tool")
        if len(set(descriptor.tools)) != len(descriptor.tools):
            raise BridgeError("registry_collision", "Bridge capability module declares a duplicate tool name")
        for tool in descriptor.tools:
            if not tool:
                raise BridgeError("registry_invalid", "Bridge public tool name cannot be empty")
            if tool in self._tool_owner:
                owner = self._tool_owner[tool]
                raise BridgeError(
                    "registry_collision",
                    f"Bridge public tool name is duplicated by modules {owner!r} and {descriptor.module_id!r}",
                )
        self._modules[descriptor.module_id] = descriptor
        self._tool_owner.update({tool: descriptor.module_id for tool in descriptor.tools})

    def _selected_module_ids(self) -> tuple[str, ...]:
        if self._enabled_selector is None:
            return tuple(self._modules)
        unknown = [module_id for module_id in self._enabled_selector if module_id not in self._modules]
        if unknown:
            raise BridgeError("unknown_module", f"Unknown Bridge module selector: {', '.join(unknown)}")
        selected = set(self._enabled_selector)
        return tuple(module_id for module_id in self._modules if module_id in selected)

    def is_enabled(self, module_id: str) -> bool:
        return module_id in set(self._selected_module_ids())

    def enabled_module_ids(self) -> tuple[str, ...]:
        return self._selected_module_ids()

    def disabled_module_ids(self) -> tuple[str, ...]:
        enabled = set(self._selected_module_ids())
        return tuple(module_id for module_id in self._modules if module_id not in enabled)

    def expected_tools(self) -> tuple[str, ...]:
        enabled = set(self._selected_module_ids())
        return tuple(
            tool
            for module_id, descriptor in self._modules.items()
            if module_id in enabled
            for tool in descriptor.tools
        )

    def mount(self, server: Any) -> None:
        """Install enabled modules and fail closed on missing/orphaned tool registrations."""

        selected = set(self._selected_module_ids())
        for module_id, descriptor in self._modules.items():
            if module_id not in selected:
                continue
            before = set(server._tool_manager._tools)
            descriptor.install(server)
            after = set(server._tool_manager._tools)
            added = after - before
            declared = set(descriptor.tools)
            if added != declared:
                missing = sorted(declared - added)
                orphan = sorted(added - declared)
                raise BridgeError(
                    "registry_surface_mismatch",
                    f"Module {module_id!r} tool registration mismatch; missing={missing}, orphan={orphan}",
                )

        self._tool_source = lambda: set(server._tool_manager._tools)
        audit = self.surface_audit()
        if not audit["consistent"]:
            raise BridgeError(
                "registry_surface_mismatch",
                f"Bridge tool surface has residue; missing={audit['missing_tools']}, orphan={audit['orphan_tools']}",
            )

    def manifest(self) -> list[dict[str, Any]]:
        enabled = set(self._selected_module_ids())
        return [
            {
                "module": item.module_id,
                "version": item.version,
                "enabled": item.module_id in enabled,
                "tools": list(item.tools),
            }
            for item in self._modules.values()
        ]

    def health(self) -> list[dict[str, Any]]:
        enabled = set(self._selected_module_ids())
        results: list[dict[str, Any]] = []
        for descriptor in self._modules.values():
            if descriptor.module_id not in enabled:
                continue
            try:
                health = descriptor.health()
            except Exception:
                health = {"ok": False, "error": {"code": "health_unavailable", "message": "Capability health check failed"}}
            results.append({"module": descriptor.module_id, "version": descriptor.version, "health": health})
        return results

    def surface_audit(self) -> dict[str, Any]:
        expected = set(self.expected_tools())
        actual = expected if self._tool_source is None else set(self._tool_source())
        missing = sorted(expected - actual)
        orphan = sorted(actual - expected)
        fingerprint_input = "\n".join(sorted(actual)).encode("utf-8")
        return {
            "enabled_modules": list(self.enabled_module_ids()),
            "disabled_modules": list(self.disabled_module_ids()),
            "module_count": len(self.enabled_module_ids()),
            "tool_count": len(actual),
            "expected_tool_count": len(expected),
            "missing_tools": missing,
            "orphan_tools": orphan,
            "consistent": not missing and not orphan,
            "surface_fingerprint": hashlib.sha256(fingerprint_input).hexdigest(),
        }

    def shutdown(self) -> None:
        enabled = set(self._selected_module_ids())
        for descriptor in reversed(tuple(self._modules.values())):
            if descriptor.module_id not in enabled or descriptor.shutdown is None:
                continue
            try:
                descriptor.shutdown()
            except Exception:
                continue

File: capabilities/registry.go
package capabilities

import (
	"crypto/sha256"
	"fmt"
	"sort"
	"strings"
)

type HealthFunc func() map[string]any

type Module struct {
	ID           string
	Kind         string
	Version      string
	Dependencies []string
	Traits       []string
	Tools        []string
	Health       HealthFunc
}

type Registry struct {
	modules  []Module
	selected []string
}

func New(selected []string) *Registry {
	return &Registry{selected: append([]string(nil), selected...)}
}

func (r *Registry) Register(module Module) error {
	if module.ID == "" {
		return fmt.Errorf("registry_invalid: Bridge capability module ID cannot be empty")
	}
	if len(module.Tools) == 0 {
		return fmt.Errorf("registry_invalid: Bridge capability module must declare at least one public tool")
	}
	local := map[string]bool{}
	for _, tool := range module.Tools {
		if tool == "" {
			return fmt.Errorf("registry_invalid: Bridge public tool name cannot be empty")
		}
		if local[tool] {
			return fmt.Errorf("registry_collision: Bridge capability module declares a duplicate tool name")
		}
		local[tool] = true
	}
	for _, existing := range r.modules {
		if existing.ID == module.ID {
			return fmt.Errorf("registry_collision: Bridge capability module is duplicated")
		}
		owned := map[string]bool{}
		for _, tool := range existing.Tools {
			owned[tool] = true
		}
		for _, tool := range module.Tools {
			if owned[tool] {
				return fmt.Errorf("registry_collision: Bridge public tool name is duplicated by modules %q and %q", existing.ID, module.ID)
			}
		}
	}
	r.modules = append(r.modules, Module{
		ID: module.ID, Kind: module.Kind, Version: module.Version,
		Dependencies: append([]string(nil), module.Dependencies...),
		Traits:       append([]string(nil), module.Traits...),
		Tools:        append([]string(nil), module.Tools...), Health: module.Health,
	})
	return nil
}

func (r *Registry) ValidateSelection() error {
	known := map[string]bool{}
	for _, module := range r.modules {
		known[module.ID] = true
	}
	for _, id := range r.selected {
		if !known[id] {
			return fmt.Errorf("unknown_module: Unknown Bridge module selector: %s", id)
		}
	}
	for _, module := range r.modules {
		for _, dependency := range module.Dependencies {
			if !known[dependency] {
				return fmt.Errorf("registry_invalid: Bridge module %s references unknown dependency %s", module.ID, dependency)
			}
			if r.IsEnabled(module.ID) && !r.IsEnabled(dependency) {
				return fmt.Errorf("module_dependency_missing: %s requires %s", module.ID, dependency)
			}
		}
	}
	return nil
}

func (r *Registry) IsEnabled(moduleID string) bool {
	if len(r.selected) == 0 {
		for _, module := range r.modules {
			if module.ID == moduleID {
				return true
			}
		}
		return false
	}
	for _, id := range r.selected {
		if id == moduleID {
			return true
		}
	}
	return false
}

func (r *Registry) Enabled() []Module {
	out := []Module{}
	for _, module := range r.modules {
		if r.IsEnabled(module.ID) {
			out = append(out, module)
		}
	}
	return out
}

func (r *Registry) EnabledIDs() []string {
	out := []string{}
	for _, module := range r.Enabled() {
		out = append(out, module.ID)
	}
	return out
}

func (r *Registry) DisabledIDs() []string {
	out := []string{}
	for _, module := range r.modules {
		if !r.IsEnabled(module.ID) {
			out = append(out, module.ID)
		}
	}
	return out
}

func (r *Registry) ExpectedTools() []string {
	out := []string{}
	for _, module := range r.Enabled() {
		out = append(out, module.Tools...)
	}
	return out
}

func (r *Registry) Manifest() []map[string]any {
	out := make([]map[string]any, 0, len(r.modules))
	for _, module := range r.modules {
		out = append(out, map[string]any{
			"module": module.ID, "kind": module.Kind, "version": module.Version, "enabled": r.IsEnabled(module.ID),
			"dependencies": append([]string(nil), module.Dependencies...), "traits": append([]string(nil), module.Traits...),
			"tools": append([]string(nil), module.Tools...),
		})
	}
	return out
}

func (r *Registry) Health() []map[string]any {
	out := []map[string]any{}
	for _, module := range r.Enabled() {
		health := map[string]any{"ok": true}
		if module.Health != nil {
			func() {
				defer func() {
					if recover() != nil {
						health = map[string]any{"ok": false, "error": map[string]any{"code": "health_unavailable", "message": "Capability health check failed"}}
					}
				}()
				health = module.Health()
			}()
		}
		out = append(out, map[string]any{"module": module.ID, "version": module.Version, "health": health})
	}
	return out
}

func (r *Registry) SurfaceAudit(actual []string) map[string]any {
	expectedSet := map[string]bool{}
	for _, tool := range r.ExpectedTools() {
		expectedSet[tool] = true
	}
	actualSet := map[string]bool{}
	for _, tool := range actual {
		actualSet[tool] = true
	}
	missing := []string{}
	orphan := []string{}
	for tool := range expectedSet {
		if !actualSet[tool] {
			missing = append(missing, tool)
		}
	}
	for tool := range actualSet {
		if !expectedSet[tool] {
			orphan = append(orphan, tool)
		}
	}
	sort.Strings(missing)
	sort.Strings(orphan)
	sortedActual := make([]string, 0, len(actualSet))
	for tool := range actualSet {
		sortedActual = append(sortedActual, tool)
	}
	sort.Strings(sortedActual)
	sum := sha256.Sum256([]byte(strings.Join(sortedActual, "\n")))
	return map[string]any{
		"enabled_modules": r.EnabledIDs(), "disabled_modules": r.DisabledIDs(),
		"module_count": len(r.Enabled()), "tool_count": len(actualSet), "expected_tool_count": len(expectedSet),
		"missing_tools": missing, "orphan_tools": orphan, "consistent": len(missing) == 0 && len(orphan) == 0,
		"surface_fingerprint": fmt.Sprintf("%x", sum),
	}
}
